src/utils/permissions.ts

```diff
@@ -1,117 +1,38 @@
-}
-
-/** Checks if the member has this permission. If the member is an owner or has admin perms it will always be true. */
-export function memberHasPermission(
-  memberID: string,
-  guild: Guild,
-  memberRoleIDs: string[],
-  permissions: Permission[],
-) {
-  if (memberID === guild.ownerID) return true;
-
-  const permissionBits = memberRoleIDs.map((id) =>
-    guild.roles.get(id)?.permissions
-  )
-    // Removes any edge case undefined
-    .filter((id) => id)
-    .reduce((bits, permissions) => {
-      bits |= BigInt(permissions);
-      return bits;
-    }, BigInt(0));
-
-  if (permissionBits & BigInt(Permissions.ADMINISTRATOR)) return true;
-
-  return permissions.every((permission) =>
-    permissionBits & BigInt(Permissions[permission])
-  );
-}
-
-export async function botHasPermission(
-  guildID: string,
-  permissions: Permission[],
-) {
-  const guild = await cacheHandlers.get("guilds", guildID);
-  if (!guild) return false;
-
-  // Check if the bot is the owner of the guild, if it is, returns true
-  if (guild.ownerID === botID) return true;
-
-  const member = (await cacheHandlers.get("members", botID))?.guilds.get(
-    guildID,
-  );
-  if (!member) return false;
-
-  // The everyone role is not in member.roles
-  const permissionBits = [...member.roles, guild.id]
-    .map((id) => guild.roles.get(id)!)
-    // Remove any edge case undefined
-    .filter((r) => r)
-    .reduce((bits, data) => {
-      bits |= BigInt(data.permissions);
-
-      return bits;
-    }, BigInt(0));
-
-  if (permissionBits & BigInt(Permissions.ADMINISTRATOR)) return true;
-
-  return permissions.every((permission) =>
-    permissionBits & BigInt(Permissions[permission])
-  );
-}
-
-/** Checks if the bot has the permissions in a channel */
-export function botHasChannelPermissions(
-  channelID: string,
-  permissions: Permission[],
-) {
-  return hasChannelPermissions(channelID, botID, permissions);
-}
-
-/** Checks if a user has permissions in a channel. */
-export async function hasChannelPermissions(
+import { Role } from "../structures/structures.ts";
+import {
+  Errors,
+  Permission,
+  Permissions,
+  RawOverwrite,
+} from "../types/types.ts";
+
+export async function calculateChannelPermissions(
   channelID: string,
   memberID: string,
   permissions: Permission[],
 ) {
   const channel = await cacheHandlers.get("channels", channelID);
-  if (!channel) return false;
+  if (!channel) return "CHANNEL_NOT_FOUND";
   if (!channel.guildID) return true;
 
   const guild = await cacheHandlers.get("guilds", channel.guildID);
-  if (!guild) return false;
+  if (!guild) return "GUILD_NOT_FOUND";
 
   if (guild.ownerID === memberID) return true;
-  if (
-    await memberIDHasPermission(memberID, guild.id, ["ADMINISTRATOR"])
-  ) {
-    return true;
-  }
+
+  const isAdmin = await calculateServerPermissions(
+    memberID,
+    guild.id,
+    ["ADMINISTRATOR"],
+  );
+  if (isAdmin === true) return true;
```

Comment:
This for sure needs work. Don't like this implementation

Reply:
How do you want to change that?